main.cpp

```diff
@@ -0,0 +1,306 @@
+#include <iostream>
+#include <fstream>
+template <typename T>
+class Matrix {
+private:
+    int rows, cols;
+    T** data;
+public:
+    Matrix() : rows(0), cols(0), data(nullptr) {}
+    Matrix(int rows_t, int cols_t, T** data2) {
+        rows = rows_t;
+        cols = cols_t;
+        data = new T*[rows];
+        for (int i = 0; i < rows; ++i) {
+            data[i] = new T[cols];
+            for (int j = 0; j < cols; ++j) {
+                data[i][j] = data2[i][j];
+            }
+        }
+    }
+    Matrix(const Matrix& other)  {
+        rows = other.rows;
+        cols = other.cols;
+        data = new T*[rows];
+        for (int i = 0; i < rows; ++i) {
+            data[i] = new T[cols];
+            for (int j = 0; j < cols; ++j) {
+                data[i][j] = other.data[i][j];
+            }
+        }
+    }
+    ~Matrix() {
+        if (rows > 0){
+            for (int i = 0; i < rows; ++i) {
+                delete[] data[i];
+            }
+            delete[] data;
+        }
+    }
+
+    Matrix& operator=(const Matrix& other) {
+        if (&other == this) {
+            return *this;
+        }
+        if (rows > 0){
+            for (int i = 0; i < rows; ++i) {
+                delete[] data[i];
+            }
+            delete[] data;
+        }
+        delete[] data;
+        rows = other.rows;
+        cols = other.cols;
+        data = new T*[rows];
+        for (int i = 0; i < rows; ++i) {
+            data[i] = new T[cols];
+            for (int j = 0; j < cols; ++j) {
+                data[i][j] = other.data[i][j];
+            }
+        }
+        return *this;
+    }
+    T* operator[](int row) {
+        return data[row];
+    }
+    Matrix operator*(const Matrix& other) const {
+        if (cols != other.rows) {
+            throw std::invalid_argument("Matrices have incompatible dimensions");
+        }
+        T** temp_ar;
+        temp_ar = new T*[rows];
+        for (int i = 0; i < rows;++i){

```

Comment:
Memory allocation for this array can be taken out in another function because it repeats later.